elastic/elastic-agent
# Enhance mock fleet-server and add --proxy-url integration test

## What does this PR do?

### Extends `testing/fleetservertest` to work out of the box with an elastic-agent and a fake component:

See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it all the enhancements.

Use cases:

#### test simple agent <-> fleet-serve interactions: Done
A single fleet-server handler involved.
##### Minimal requirements:
 - [x] easy to implement fleet-server handler
 - [x] run it as a valid HTTP server
Nice  have: New Handler functions taking  a few options to tailor the handler for
each test. Check [`handlers.go`](handlers.go) for the available paths and handlers.

#### Run the agent against the mock fleet-server
Allow to e2e test the Agent without using a real fleet-server which allows:
 - test error cases / unhappy paths / hard to reproduce states. Eg. fleet-server sends an invalid policy
 - develop and test features before they're fully implemented on fleet-server

in order run the agent against the mock fleet-server, it needs to:
 - [x] have the states, enroll, checkin and ack handlers implemented and "working together":
   - [x]  it enrols one agent
   - [x]  the checkins:
     - [x]  are for the enrolled agent
     - [x]  might be set to fail
     - [x]  actions can be added and later sent to the agent
     - [x]  nice to have:
       - [x]  choose how long the checkin handler takes to send the repose (simulate long poll)
   - [x]  track acks:
     - [x]  once an action is acked, it won't be sent again
     - [x]  acks can be set to fail
 - ~~[ ]  allowed integrations (no need for real ES):~~
     - ~~[ ]  fake component~~
     - ~~[ ]  fake shipper~~
 - [x]  an API key and Enrollment token: both ignored, there is no authentication

### Add `proxytest.Proxy`, a naive proxy intended for tests which allows to configure URL rewrite and check all the requests made to it.
See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it.

### Add E2E tests for fleet-server's proxy:

 - [x] `--proxy-url` set and no proxy in the policy
 - [x] `--proxy-url` set and empty proxy in the policy
 - [x] `--proxy-url` set and valid proxy in the policy
 - [x] no `--proxy-url` set and valid proxy in the policy
 - [x] proxy set, the remove proxy from policy


## Why is it important?

Enable us to test better agent->fleet-server interactions without the need to deploy the whole stack as well as to test unusual corner cases which are had to reproduce with a real fleet-server.

Also the tests for the agent's fleet-server proxy ~should~ will ensure we do not introduce any regression.

## Checklist




## How to test this PR locally

See [proxy_url_test.go](testing/integration/proxy_url_test.go) for examples on how to use it all the mock fleet-server enhancements and the `proxytest.Proxy`.

### To use the mock fleet server
 - Comment out the test skip of `TestRunFleetServer`
 - run the test: `go test -v -run TestRunFleetServer ./testing/fleetservertest`
 - use the shown server address as the `--url` enroll parameter and any value for `--enrollment-token`
 - the agent successfully enroll.

- run the server:
```
❯ go test -v -run TestRunFleetServer ./testing/fleetservertest
=== RUN   TestRunFleetServer
listening on: "http://[::]:42833"
press CTRL + C or kill the test to stop
```

- enroll the agent
```
root@elastic-agent-dev:~# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
```

### To run the integration tests:

 - remotely:
```
GOTEST_FLAGS="-test.run TestProxyURL" mage integration:test
```

 - locally: **be warned it should be done on a VM as it'll replace any installed agent**
```
vagrant up elastic-agent
vagrant ssh elastic-agent

cd /vagrant
TEST_DEFINE_PREFIX=gambiarra go test -run TestProxyURL -v -tags integration -timeout 0 ./testing/integration/
```

## Related issues

- Relates to #2482


## Logs

 - agent vagrant VM
```
root@elastic-agent-dev:~# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
sudo: ./elastic-agent: command not found
root@elastic-agent-dev:~# cd ./elastic-agent-8.10.0-SNAPSHOT-linux-x86_64/
root@elastic-agent-dev:~/elastic-agent-8.10.0-SNAPSHOT-linux-x86_64# sudo ./elastic-agent install --url=http://192.168.56.1:42833 --enrollment-token=anythingWIllDO -nif
Installing in non-interactive mode.{"log.level":"warn","@timestamp":"2023-07-03T17:35:27.908Z","log.logger":"tls","log.origin":{"file.name":"tlscommon/tls_config.go","file.line":104},"message":"SSL/TLS verifications disabled.","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-07-03T17:35:27.919Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":478},"message":"Starting enrollment to URL: http://192.168.56.1:42833/","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-07-03T17:35:29.366Z","log.origin":{"file.name":"cmd/enroll_cmd.go","file.line":276},"message":"Successfully triggered restart on running Elastic Agent.","ecs.version":"1.6.0"}
Successfully enrolled the Elastic Agent.
Elastic Agent has been successfully installed.
root@elastic-agent-dev:~/elastic-agent-8.10.0-SNAPSHOT-linux-x86_64# 
```

 - fleet server mock
```
❯ go test -v -run TestRunFleetServer ./testing/fleetservertest
=== RUN   TestRunFleetServer
listening on: "http://[::]:42833"
press CTRL + C or kill the test to stop
    server.go:127: 200 - POST /api/fleet/agents/enroll? HTTP/1.1 192.168.56.42:36772
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162
    server.go:127: 200 - POST /api/fleet/agents/agentID/checkin? HTTP/1.1 192.168.56.42:49162

```


## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
